modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/TxManagerImpl.java

```diff
@@ -295,139 +319,252 @@ public void finishFull(HybridTimestampTracker timestampTracker, UUID txId, boole
+        AtomicBoolean performingFinish = new AtomicBoolean();
         TxContext tuple = txCtxMap.compute(txId, (uuid, tuple0) -> {
             if (tuple0 == null) {
                 tuple0 = new TxContext(); // No writes enlisted.
             }
 
-            if (tuple0.finishFut == null) {
-                tuple0.finishFut = new CompletableFuture<>();
-                ref.set(tuple0.finishFut);
+            if (!tuple0.isTxFinishing()) {
+                tuple0.finishTx();
+
+                performingFinish.set(true);
             }
 
             return tuple0;
         });
 
-        if (ref.get() != null) { // This is a finishing thread.
-            if (!commit) {
-                clo.apply(null).handle((ignored, err) -> {
+        // This is a finishing thread.
+        if (performingFinish.get()) {
+            Function<Void, CompletableFuture<Void>> finishAction = ignored ->
+                    makeFinishRequest(
+                            observableTimestampTracker,
+                            commitPartition,
+                            commit,
+                            enlistedGroups,
+                            txId,
+                            finishingStateMeta.txFinishFuture()
+                    );
+
+            runFinish(commit, tuple, finishAction);
+        }
+
+        // The method `runFinish` has a side effect on `finishInProgressFuture` future, it kicks off another future that will complete it.
+        return tuple.finishInProgressFuture;
+    }
+
+    private static void runFinish(boolean commit, TxContext tuple, Function<Void, CompletableFuture<Void>> finishAction) {
+        // Wait for commit acks first, then proceed with the finish request.
+        CompletableFuture<Void> finisher = commit ? tuple.waitNoInflights() : completedFuture(null);
+
+        finisher
+                .thenCompose(finishAction)
+                .handle((ignored, err) -> {
                     if (err == null) {
-                        tuple.finishFut.complete(null);
+                        tuple.finishInProgressFuture.complete(null);
                     } else {
-                        tuple.finishFut.completeExceptionally(err);
+                        tuple.finishInProgressFuture.completeExceptionally(err);
                     }
                     return null;
                 });
-            } else {
+    }
 
-                // All inflights have been completed before the finish.
-                if (tuple.inflights == 0) {
-                    tuple.waitRepFut.complete(null);
-                }
+    private CompletableFuture<Void> makeFinishRequest(
+            HybridTimestampTracker observableTimestampTracker,
+            TablePartitionId commitPartition,
+            boolean commit,
+            Map<TablePartitionId, Long> enlistedGroups,
+            UUID txId,
+            CompletableFuture<TransactionMeta> txFinishFuture
+    ) {
+        HybridTimestamp commitTimestamp = getCommitTimestamp(commit);
```

Comment:
I would suggest to leave the generation of commit timestamp in `finish` method to emphasize that it must be generated only once. It's not obvious in this method and can be lost during some refactoring. But I also suggest to refactor it, as I said in another comment.